apps/emqx/src/emqx_ds_schema.erl

```diff
@@ -0,0 +1,245 @@
+
+translate_builtin(#{
+    backend := builtin,
+    n_shards := NShards,
+    replication_factor := ReplFactor,
+    layout := Layout
+}) ->
+    Storage =
+        case Layout of
+            #{
+                type := wildcard_optimized,
+                bits_per_topic_level := BitsPerTopicLevel,
+                epoch_bits := EpochBits,
+                topic_index_bytes := TIBytes
+            } ->
+                {emqx_ds_storage_bitfield_lts, #{
+                    bits_per_topic_level => BitsPerTopicLevel,
+                    topic_index_bytes => TIBytes,
+                    epoch_bits => EpochBits
+                }};
+            #{type := reference} ->
+                {emqx_ds_storage_reference, #{}}
+        end,
+    #{
+        backend => builtin,
+        n_shards => NShards,
+        replication_factor => ReplFactor,
+        storage => Storage
+    }.
+
+%%================================================================================
+%% Behavior callbacks
+%%================================================================================
+
+namespace() ->
+    durable_storage.
+
+schema() ->
+    [
+        {"messages",
+            ds_schema(#{
+                desc => ?DESC(messages),
+                importance => ?IMPORTANCE_HIDDEN,
+                default =>
+                    #{
+                        <<"backend">> => builtin
+                    }
+            })}
+    ].
+
+fields("builtin") ->
+    %% Schema for the builtin backend:
+    [
+        {"backend",
+            sc(
+                builtin,
+                #{
+                    importance => ?IMPORTANCE_MEDIUM,
+                    'readOnly' => true,
+                    default => builtin,
+                    desc => ?DESC(builtin)
+                }
+            )},
+        {"_config_handler",
+            sc(
+                {module(), atom()},
+                #{
+                    importance => ?IMPORTANCE_HIDDEN,
+                    'readOnly' => true,
+                    default => {?MODULE, translate_builtin}
+                }
+            )},
+        {"data_dir",
+            sc(
+                string(),
+                #{
+                    desc => ?DESC(builtin_data_dir),
+                    mapping => "emqx_durable_storage.db_data_dir",
+                    required => false,
+                    importance => ?IMPORTANCE_MEDIUM
+                }
+            )},
+        {"n_shards",
+            sc(
+                pos_integer(),
+                #{
+                    importance => ?IMPORTANCE_MEDIUM,
+                    desc => ?DESC(builtin_n_shards),
+                    default => 16
+                }
+            )},
+        {"replication_factor",
+            sc(
+                pos_integer(),
+                #{
+                    default => 3,
+                    importance => ?IMPORTANCE_HIDDEN
+                }
+            )},
+        {"egress",
```

Comment:
not sure if 'egress' is the best name here, any alternatives?

Reply:
I guess it can be just named "buffer", but there are just too many different buffers.